datafusion/optimizer/src/handle_gap_fill.rs

```diff
@@ -0,0 +1,451 @@
+};
+use std::sync::Arc;
+
+/// This optimizer rule enables gep-filling semantics for SQL queries
+/// that contain calls to `DATE_BIN_GAPFILL()`.
+///
+/// In SQL a typical gap-filling query might look like this:
+/// ```sql
+/// SELECT
+///   location,
+///   DATE_BIN_GAPFILL(INTERVAL '1 minute', time, '1970-01-01T00:00:00Z') AS minute,
+///   AVG(temp)
+/// FROM temps
+/// WHERE time > NOW() - INTERVAL '6 hours' AND time < NOW()
+/// GROUP BY LOCATION, MINUTE
+/// ```
+/// The aggregateion step of the initial logical plan looks like this:
+/// ```text
+///   Aggregate: groupBy=[[datebingapfill(IntervalDayTime("60000"), temps.time, TimestampNanosecond(0, None)))]], aggr=[[AVG(temps.temp)]]
+/// ```
+/// However, `DATE_BIN_GAPFILL()` does not have an actual implementation like other functions.
+/// Instead, the plan is transformed to this:
+/// ```text
+/// GapFill: groupBy=[[datebingapfill(IntervalDayTime("60000"), temps.time, TimestampNanosecond(0, None)))]], aggr=[[AVG(temps.temp)]], start=..., stop=...
+///   Sort: datebingapfill(IntervalDayTime("60000"), temps.time, TimestampNanosecond(0, None))
+///     Aggregate: groupBy=[[datebingapfill(IntervalDayTime("60000"), temps.time, TimestampNanosecond(0, None)))]], aggr=[[AVG(temps.temp)]]
+/// ```
+/// This optimizer rule makes that transformation.
+pub struct HandleGapFill;
+
+impl HandleGapFill {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl Default for HandleGapFill {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl OptimizerRule for HandleGapFill {
+    fn try_optimize(
+        &self,
+        plan: &LogicalPlan,
+        _config: &dyn OptimizerConfig,
+    ) -> Result<Option<LogicalPlan>> {
+        handle_gap_fill(plan)
+    }
+
+    fn name(&self) -> &str {
+        "handle_gap_fill"
+    }
+
+    fn apply_order(&self) -> Option<ApplyOrder> {
+        Some(ApplyOrder::BottomUp)
+    }
+}
+
+fn handle_gap_fill(plan: &LogicalPlan) -> Result<Option<LogicalPlan>> {
+    let res = match plan {
+        LogicalPlan::Aggregate(_) => handle_aggregate(plan)?,
+        _ => None,
+    };
+
+    if res.is_none() {
+        // no transformation was applied,
+        // so make sure the plan is not using gap filling
+        // functions in an unsupported way.
+        check_node(plan)?;
+    }
+
+    Ok(res)
+}
+
+fn handle_aggregate(plan: &LogicalPlan) -> Result<Option<LogicalPlan>> {
+    let (group_expr, aggr_expr, input) = match &plan {
+        LogicalPlan::Aggregate(Aggregate {
+            group_expr,
+            aggr_expr,
+            input,
+            ..
+        }) => (group_expr, aggr_expr, input),
+        _ => unreachable!(),
+    };
+
+    let new_group_expr = replace_date_bin_gapfill(group_expr)?;
+    if new_group_expr.is_none() {
+        // nothing to do
+        return Ok(None);
+    }
+    // new_group_expr has DATE_BIN_GAPFILL replaced with DATE_BIN.
+    let (new_group_expr, dbg_idx) = new_group_expr.unwrap();
```

Comment:
another possible more "rusty" style for this (to avoid the unwrap) is:


```suggestion
    let new_group_expr = replace_date_bin_gapfill(group_expr)?;
    if let Some(new_group_expr) =  replace_date_bin_gapfill(group_expr)? {
      new_group_expr
    } else {
        // nothing to do
        return Ok(None);
    };
    let (new_group_expr, dbg_idx) = new_group_expr;
```